File: Mobile/Server2/app/lib/main.dart
import 'dart:async';
import 'dart:convert';

import 'package:app/server/ApiRequests.dart';
import 'package:app/view/EditService.dart';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:app/database/database_service.dart';
import 'package:app/model/Service.dart';
import 'package:app/view/AddService.dart';
import 'package:app/view/ServiceNotifier.dart';

Future<void> syncCachedRequests() async {
  try {
    final cachedRequests = await DatabaseService.instance.getCachedRequests();

    for (var cachedRequest in cachedRequests) {
       final method = cachedRequest['method'];
        final bodyString = cachedRequest['body'];

      // Parse the JSON string into a Dart map
        final body = json.decode(bodyString);
        
        print('Body for cached request is: $body');
        final id = body['id'];
    
      try {
        if (method.toLowerCase() == 'put') {
          print("Sending PUT request for cached service:$id");
          // If the method is PUT, assume it's an 'update' request
          await ApiRequests.putRequest('/update', body).timeout(
            Duration(seconds: 2),
            onTimeout: () {
              // Handle the timeout
              print('Timeout syncing cached request: $id');
            },
          );
        } else if (method.toLowerCase() == 'delete') {
          
          print("Sending DELETE request for cached service:$id");
          // If the method is DELETE, assume it's a 'delete' request
          await ApiRequests.deleteRequest('/delete/$id').timeout(
            Duration(seconds: 2),
            onTimeout: () {
              // Handle the timeout
              throw TimeoutException('The request to the server timed out');
            },
          );
        }

        // Remove the cached request as it was successfully sent to the server
        await DatabaseService.instance.removeCachedRequest(cachedRequest['id']);
      } catch (e) {
        // Log the error message to the console
        print('Error syncing cached request: $e');
      }
    }
  } catch (e) {
    // Log the error message to the console
    print('Error syncing cached requests: $e');
  }
}

Future<void> syncLocalServicesToServer(List<Service> localServices) async {

  // Post services that are fresh (negative ID)
  for (var service in localServices) {
    if (service.id < 0) {
      // Service has a negative ID, indicating it is not yet synced with the server
      try {
        // Post to server
        await ApiRequests.postRequest(
          '/add_service',
          service.toMap(),
        ).timeout(
          Duration(seconds: 2),
          onTimeout: () {
            // Handle the timeout
            print('Timeout syncing service to the server:${service.id}');
          },
        );

        
      } catch (e) {
        // Handle the error
        print('Error posting service ${service.id} to the server: $e');
      }
    }
  }

  // Make updates/ deletes on already existing services in server
   await syncCachedRequests();
}


void main() async {
  WidgetsFlutterBinding.ensureInitialized();

  
  //  Check for an internet connection
  bool isInternetConnected = await ApiRequests.isInternetConnected();

  List<Service> services = [];

  //  If there is an internet connection, fetch data from the server
  if (isInternetConnected) {
    
    // Get local DB Services
      List<Service> localDbServices = await DatabaseService.instance.getServices();


      // Sync local services with negative IDs to the server
      await syncLocalServicesToServer(localDbServices);

      
    try {
      print("Trying to fetch data from server");

      // Use the timeout feature to set a 5-second timeout for the network request

      List<Service> serverServices = await ApiRequests.getServices().timeout(
      Duration(seconds: 10),
      onTimeout: () {
        // This callback is called if the network request times out
        throw TimeoutException('The request to the server timed out');
      },
    ).then((serviceMap) {
      return serviceMap.map((serviceMap) => Service.fromMap(serviceMap)).toList();
    });

      // Clear the local database after successfully syncing with the server
      await DatabaseService.instance.clearLocalDatabase();
  
      
      // Insert services received from serer
      await DatabaseService.instance.insertServices(serverServices); 

      services = serverServices;
    } catch (e) {
      print("Error fetching services from the server: $e");

      // If fetching from the server fails, fetch data from the local database
      services = await DatabaseService.instance.getServices();
    }
  } else {
    // If there is no internet connection, fetch data from the local database
    services = await DatabaseService.instance.getServices();
  }

  runApp(
    MultiProvider(
      providers: [
        ChangeNotifierProvider(create: (context) => ServiceNotifier()..setServices(services)),
      ],
      child: const MyApp(),
    ),
  );
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    
    return MaterialApp(
      title: 'Cook Book',
      theme: ThemeData(
       
    
    
      ),
      home: const MyHomePage(title: 'Services Page'),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(title),
        backgroundColor: Color.fromARGB(255, 0, 195, 10),
      ),
      backgroundColor: Color.fromARGB(255, 0, 201, 191),
      body: Consumer<ServiceNotifier>(
        builder: (context, serviceNotifier, child) {
          final services = serviceNotifier.services;
          return ServiceList(services: services);
        },
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () async {
          final newService = await Navigator.of(context).push(
            MaterialPageRoute(
              builder: (context) => AddService(),
            ),
          );

          if (newService != null) {
            // Add the new service to the list
            context.read<ServiceNotifier>().addService(newService);
            final scaffoldMessenger = ScaffoldMessenger.of(context);
            scaffoldMessenger.showSnackBar(
              SnackBar(
                content: Text('New service added successfully'),
                duration: Duration(seconds: 2), 
              ),
            );
          }
        },
        tooltip: 'Add Service',
        backgroundColor: Color(0xFF023047),
        child: Icon(Icons.add),
      ),
    );
  }
}

class ServiceList extends StatelessWidget {
  final List<Service> services;

  const ServiceList({Key? key, required this.services}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ListView.builder(
      itemCount: services.length,
      itemBuilder: (context, index) {
        final service = services[index];
        return GestureDetector(
          onTap: () async {
            
            final updatedService = await Navigator.of(context).push(
              MaterialPageRoute(
              
                builder: (context) => EditService(service: service),
              ),
            );
          
            if (updatedService != null) {
             if (updatedService is Service) {
                    // The service was updated, we returned a Service type
                    // Update the service in the UI
                    context.read<ServiceNotifier>().updateService(updatedService);
                    final scaffoldMessenger = ScaffoldMessenger.of(context);
                    scaffoldMessenger.showSnackBar(
                      SnackBar(
                        content: Text('Service updated successfully'),
                        duration: Duration(seconds: 2), 
                      ),
                    );
                  }
                  else
                      if(updatedService is int) {
                        
                        // The service was deleted, we returned an int - the ID of the deleted service
                        final scaffoldMessenger = ScaffoldMessenger.of(context);
                        scaffoldMessenger.showSnackBar(
                          SnackBar(
                            content: Text('Service deleted successfully'),
                            duration: Duration(seconds: 2), 
                          ),
                        );
                        
                        context.read<ServiceNotifier>().deleteService(updatedService);
                       
                      }
                }
          },
          child: ListTile(
  title: Text('${service.name}',  style: TextStyle(fontSize: 20,) ),
 
  subtitle: Padding(
    padding: const EdgeInsets.all(8.0), 
    child: Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text('Provider: ${service.provider}', style: TextStyle(fontSize: 15),),
       
      ],
    ),
  ),
)
        );
      },
    );
  }
}

File: Mobile/Server2/main.py
import asyncio
import json
#import threading
from threading import Thread
from flask_cors import CORS
import websockets
from flask_sqlalchemy import SQLAlchemy
from flask import Flask, jsonify, request, render_template

app = Flask(__name__)
app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:///services.db'  # SQLite database file
db = SQLAlchemy(app)
connected_clients = set()


class Service(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(255), nullable=False)
    provider = db.Column(db.String(255), nullable=False)
    location = db.Column(db.Text, nullable=False)
    radius = db.Column(db.Integer, default=0)
    phone = db.Column(db.String(255), nullable=False)
    price = db.Column(db.Integer, nullable=False)

    def to_dict(self):
        return {
            'id': self.id,
            'name': self.name,
            'provider': self.provider,
            'location': self.location,
            'radius' : self.radius,
            'phone' : self.phone,
            'price': self.price,
        }

@app.route('/')
def home():
    return render_template('index.html', message='Hello')

@app.route('/add_service', methods=['POST'])
def add_service():
    # Print the current thread ID
    print(f"Handling request in thread {threading.current_thread().ident}")

    data = request.get_json()
    new_service = Service(
        name=data['name'],
        provider=data['provider'],
        location=data['location'],
        radius=data['radius'],
        phone=data['phone'],
        price=data['price'],
    )

    db.session.add(new_service)
    db.session.commit()
    print("Service added successfully")
    return jsonify({'message': 'Service added successfully'})

@app.route('/update', methods=['PUT'])
def update_service():
    # Print the current thread ID
    print(f"Handling request in thread {threading.current_thread().ident}")

    data = request.get_json()
    service_id = data.get('id')
    service = Service.query.get(service_id)

    if service:
        # Update the recipe with the data received in the request
        service.name = data['name']
        service.provider = data['provider']
        service.location = data['location']
        service.radius = data['radius']
        service.phone = data['phone']
        service.price = data['price']

        db.session.commit()
        print("Service updated successfully")
        return jsonify({'message': 'Service updated successfully'})
    else:
        return jsonify({'message': 'Service not found'}), 404

@app.route('/services', methods=['GET'])
def get_all_services():

    # Print the current thread ID
    print(f"Handling request in thread {threading.current_thread().ident}")

    services = Service.query.all()
    services_list = [service.to_dict() for service in services]
    return jsonify({'services': services_list})

@app.route('/delete/<int:service_id>', methods=['DELETE'])
def delete_service(service_id):
    service = Service.query.get(service_id)

    # Print the current thread ID
    print(f"Handling request in thread {threading.current_thread().ident}")

    if service:
        db.session.delete(service)
        db.session.commit()
        print("Service deleted successfully")
        return jsonify({'message': 'Service deleted successfully'})
    else:
        return jsonify({'message': 'Service not found'}), 404

async def send(websocket, message):
    try:
        await websocket.send(message)
    except websockets.ConnectionClosed:
        CLIENTS.remove(websocket)
        pass


async def broadcast(message):
    for websocket in CLIENTS:
        asyncio.create_task(send(websocket, message))


async def echo(websocket):
    CLIENTS.append(websocket)
    try:
        # data = await websocket.recv()
        # print(data)

        await websocket.wait_closed()
    finally:
        CLIENTS.remove(websocket)


async def main_sockets():
    async with websockets.serve(echo, "0.0.0.0", 8765, ssl=None):
        await asyncio.Future()


def routine1():
    asyncio.run(main_sockets(), debug=False)


def routine2():
    app.run(host="0.0.0.0", debug=False, threaded=True)



if __name__ == '__main__':
    with app.app_context():
        db.create_all()  # Create tables before running the app


    app.run(host='0.0.0.0', port=5000, debug=True, threaded=True)
    # Flash has built-in with the "threaded = True" option:
    # implementation using the SocketServer.ThreadingMixIn class
